### PR TITLE
Add Pascal VOC Class Segmentation

### PR DESCRIPTION
**Pull request by desimone**
- I use the train/val/test split defined the text files found in `./VOCdevkit/VOC2012/ImageSets/Segmentation/*.txt`
- Perhaps it makes sense to rename the class in line with its associated task like ms-coco does with `CocoDetection` and `CocoCaptions`. How about `PascalSegmentation`?
- Should any special considerations or changes be made for semantic segmentation datasets?

### CODE REVIEW

File: torchvision/datasets/pascal.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                assert os.path.isfile(image)
+                assert os.path.isfile(mask)
+                self.images.append(image)
+                self.masks.append(mask)
+
+        assert (len(self.images) == len(self.masks))
+
+    def __getitem__(self, index):
+        img = Image.open(self.images[index]).convert('RGB')
+        target = Image.open(self.masks[index]).convert('RGB')
```
Comment (fmassa) [inline torchvision/datasets/pascal.py]: This makes the target be a RGB image, and the labels become triplets. I think it's better to remove the `.convert('RGB')` to get the labels in the range 0-21, so that we can directly use them for training.

File: torchvision/datasets/pascal.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+        if self.transform is not None:
+            print("transform was not none")
+            img = self.transform(img)
+        # todo(bdd) : perhaps transformations should be applied differently to masks?
```
Comment (fmassa) [inline torchvision/datasets/pascal.py]: This is similar to https://github.com/pytorch/vision/issues/9 .
I think we will need to change the way we apply transforms to allow randomly transforming both the input and the target at the same time.
For the time being, I think this is fine like this.

File: torchvision/datasets/pascal.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                pass
+            else:
+                raise
+
+        if self._check_integrity():
+            print('Files already downloaded and verified')
+            return
+
+        # downloads file
+        if os.path.isfile(fpath) and \
```
Comment (fmassa) [inline torchvision/datasets/pascal.py]: Isn't this redundant with the `_check_integrity()` just above?

File: torchvision/datasets/pascal.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        splits_dir = os.path.join(voc_root, 'ImageSets/Segmentation')
+        split_f = os.path.join(splits_dir, 'train.txt')
+        if not self.train:
+            split_f = os.path.join(splits_dir, ' trainval.txt')
+
+        self.images = []
+        self.masks = []
+        with open(os.path.join(split_f), "r") as lines:
+            for line in lines:
+                image = os.path.join(image_dir, line.rstrip('\n') + ".jpg")
```
Comment (fmassa) [inline torchvision/datasets/pascal.py]: For small datasets like VOC, this is fine, but it might be good considering appending the `image_dir` and the `image_name` during `__getitem__` to avoid long strings where most of them are repeated content.

File: torchvision/datasets/pascal.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            self.download()
+
+        if not self._check_integrity():
+            raise RuntimeError('Dataset not found or corrupted.' +
+                               ' You can use download=True to download it')
+        # train/val/test splits are pre-cut
+        splits_dir = os.path.join(voc_root, 'ImageSets/Segmentation')
+        split_f = os.path.join(splits_dir, 'train.txt')
+        if not self.train:
+            split_f = os.path.join(splits_dir, ' trainval.txt')
```
Comment (fmassa) [inline torchvision/datasets/pascal.py]: You only consider `train' or `trainval` because `test` doesn't contain any labels, but maybe it would be good to let the user select the image set that they want to use with a string ('train', 'val', 'trainval', 'test' for example, but other splits could also be possible).
For the `test` case, maybe we could handle missing targets somehow?

**Comment by fmassa** [comment]
Thanks for the PR!
Overall this looks good. I have a few inline comments, and some more generic remarks:
* I think we should mention that this is the segmentation dataset. Maybe something like `VOCSegmentation`?
* I'd rather let the user select the split himself, instead on only allowing for `train` and `trainval` splits.
* This is specific to VOC 2012, maybe it would be good to let the user select between different year contests?
* I was planning on adding a `VOCDetection` dataset as well. Some preliminary version can be found in https://github.com/pytorch/examples/pull/21/files#diff-0344e770fabb635e92d12f8b67f3504a . I'll merge it in when this PR gets merged.

I think we will need to figure out a better way of applying random transforms to both inputs and targets, as discussed in https://github.com/pytorch/vision/issues/9 .

**Comment by desimone** [comment]
Thanks for the review, @fmassa. I have a few follow up questions while I wait for the other datasets to download...

- I've renamed the class to `VOCSegmentation`.
- Can you elaborate on how you'd want the user to select the split themselves? Should the list be built from *all* available masks and then shuffled? Sorry, I don't fully track. 
- I could add multiple years without too much trouble. I'm curious, what would the usecase be for using 2007 over say 2012?
- What would be the most idiomatic way to handle errors. At certain points I use asserts. Not sure how I feel about that.
- Common operations like archive downloading, extracting, hash checking, file io etc seem common amongst all datasets. Maybe we could abstract that functionality as an interface?
- I'll defer how to uniformly apply transforms to both mask and images per #9.

**Comment by j-min** [comment]
Any updates on this?

**Comment by fmassa** [comment]
Replaced by https://github.com/pytorch/vision/pull/663, thanks a lot for the original PR!
